leetcode/easy/958._Check_Completeness_of_a_Binary_Tree.py

```diff
@@ -0,0 +1,209 @@
+# https://leetcode.com/problems/check-completeness-of-a-binary-tree/
+
+# Definition for a binary tree node.
+# class TreeNode:
+#     def __init__(self, val=0, left=None, right=None):
+#         self.val = val
+#         self.left = left
+#         self.right = right
+
+'''
+Level order traversal:
+
+This works but its memory inefficient.
+'''
+class LevelOrder:
+    def isCompleteTree(self, root: Optional[TreeNode]) -> bool:
+        visited = []
+        queue = [(root, 0)]
```

Comment:
Again, this approach works, but it;s so much more complex than it needs to be. Try this 

1. Use a BFS (or DFS). Store a `levels = defaultdict(list)` where the key is the level and the value is the list of nodes and their left/right position in the tree. For example `node, height, order = bfs_queue.popleft()`, `bfs_queue.append(node.left, height + 1, order - 1)` Using a defaultdict(list) is much better than using a long visited list of tuples.
2. Uusing this `levels` you can check if each level is valid. 
3. Check if all levels are valid.